scratch.py

```diff
@@ -115,80 +117,96 @@ def pack_string(string):
             if not length_data:
                 if eof_is_error:
                     raise ValueError("missing or corrupt string size")
-                return 78
+                return None
 
             length = uint16.unpack(length_data)[0]
-            print(length, "2343")
-            if length == 0:
+            print(f"{length= }")
+            if not length:
                 return ""
-            data = fh.read(length)
-            print(data,"re")
-            print(data,"me",len(data),length,len(data) != length)
-            if not data or len(data) != length:
+
+            data = file.read(length)
+            print(f"{data= }")
+            print(data, len(data), length)
+            if len(data) != length:
                 raise ValueError("missing or corrupt string")
-            format = "<{0}s".format(length)
-            return struct.unpack(format, data)[0].decode("utf8")
 
-        fh = None
+            return struct.unpack(f"<{length}s", data)[0].decode("utf8")
+
+
+        NumbersStruct = struct.Struct("<Idi?")
+        file = None
         MAGIC = b"AIB\x00"
         FORMAT_VERSION = b"\x00\x01"
         GZIP_MAGIC = b"\x1F\x8B"
-        try:
-            fh = open(filename, "rb")
-            magic = fh.read(len(GZIP_MAGIC))
-            if magic == GZIP_MAGIC:
-                fh.close()
-                fh = gzip.open(filename, "rb")
-            else:
-                fh.seek(0)
-            magic = fh.read(len(MAGIC))
-            if magic != MAGIC:
-                raise ValueError("invalid .aib file format")
-            version = fh.read(len(FORMAT_VERSION))
-            if version > FORMAT_VERSION:
-                raise ValueError("unrecognized .aib file version")
-            self.clear()
-        except Exception as err:
-            print(err)
 
+        # try:
+        file = open(filename, "rb")
+        magic = file.read(len(GZIP_MAGIC))
+        if magic == GZIP_MAGIC:
+            file.close()
+            file = gzip.open(filename, "rb")
+        else:
+            file.seek(0)
+
+        magic = file.read(len(MAGIC))
+        if magic != MAGIC:
+            raise ValueError("invalid .aib file format")
+
+        version = file.read(len(FORMAT_VERSION))
+        if version > FORMAT_VERSION:
+            raise ValueError("unrecognized .aib file version")
+
+        self.clear()
+        # except Exception as err:
+        #     print(err)
+
+        data = {}
         while True:
-            report_id = unpack_string(fh, False)
-            if report_id is None:
+            report_id = unpack_string(file, False)
+            if report_id is None:  # End of data
                 break
-            data = {}
             data["report_id"] = report_id
-            for name in ("airport", "aircraft_id",
-                         "aircraft_type", "narrative"):
-                data[name] = unpack_string(fh)
-                print(data[name],34)
-                other_data = fh.read(NumbersStruct.size)
```

Comment:
**Main cause of the errror**... Due to the indentation:
* the next `NumbersStruct.size` bytes are read from the file
* causing `unpack_string()` on the next iteration to read the wrong set of bytes into `length_data`
* resulting in the expression `len(data) != length` to be **True** (when the misread length is larger than the number of bytes left in the file)
* finally raising the error.